### PR TITLE
Refactor monolithic ECS Operator into Operators, Sensors, and a Hook

### LINKED ISSUE #24826: Refactor/Review `EcsOperator`
### Body

After reading https://stackoverflow.com/questions/72630480/airflow-sensor-for-aws-ecs-task I checked the source code as it should have been relativly easy to get the functionality user is requesting but discovered that [`EcsOperator`](https://github.com/apache/airflow/blob/ec6761a5c0d031221d53ce213c0e42813606c55d/airflow/providers/amazon/aws/operators/ecs.py#L168)
seems like hook + operator + sensor in a single class which makes it harder to create a sensor as the user wishes.

For some reason we don't have `EcsHook` ? I didn't find reason why we don't have one or why the operator handle so much logic on it's own.

If we take for example:
https://github.com/apache/airflow/blob/ec6761a5c0d031221d53ce213c0e42813606c55d/airflow/providers/amazon/aws/operators/ecs.py#L420
We probably should have a function in hook that checks for status so it will be easy to extend and reuse (if for example we want to create a Sensor).

The task:
Review the class and refactor what needs to be extract to `EcsHook`, maybe create also `EcsSensor` as part of the refactor?

FYI @ferruzzi @o-nikolas @vincbeck 

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by ferruzzi**
The current EcsOperator is monolithic, does not have a hook, nor any sensors.  This PR adds an ECS hook, modifies the existing EcsOperator to use it, adds a handful of other operators and sensors.  The functionality of the existing EcsOperator was changed as little as possible.  A future refactor of  that operator might not be a bad idea, but adding the hook and sensors is the main goal of this PR.

Other changes:
 - Moved some methods from the EcsOperator module into the new EcsHook module as that felt like a better home for them.
 - Removes some old code that has been deprecated for multiple major releases.  
   - `airflow/contrib/operators/ecs_operator.py` was deprecated over 2 years ago
   - `tests/providers/amazon/aws/operators/test_ecs_system.py` is also a few years old and not a proper system test by current standards.
 -  Deprecates the name `EcsOperator` in favor of `EcsRunTaskOperator` to match the standard naming conventions since there are now multiple ECS-related operators.

closes: https://github.com/apache/airflow/issues/24826

### CODE REVIEW

**Comment by ferruzzi** [comment]
Looks like I missed removing a deprecation somewhere, looking into it.

**Comment by ferruzzi** [comment]
Sorry about that, all passing now.  :+1:

File: airflow/contrib/operators/ecs_operator.py
Comment (eladkal) [inline airflow/contrib/operators/ecs_operator.py]: This file can be removed only in Airflow 3 as its part of Airflow core so we can not remove it yet.

**Review by eladkal — DISMISSED** [review]
There are several places where code is removed.
We should deprecate first

**Comment by ferruzzi** [comment]
> There are several places where code is removed.  We should deprecate first

Unless I made a mistake, I only removed code which has been deprecated for quite some time.

File: airflow/contrib/operators/ecs_operator.py
Comment (ferruzzi) [reply airflow/contrib/operators/ecs_operator.py]: But it's so olllllldddddddd.    Git Blame shows that it was deprecated at least 2 years ago.  If it needs to stay, I'll add it back.

Comment (ferruzzi) [reply airflow/contrib/operators/ecs_operator.py]: When I revert this, what is the best way to handle it?   Should I import all of the new ECS Operators fro this PR and add them to `__all__`, or only update the names of the ones which were already included in there?

Comment (eladkal) [reply airflow/contrib/operators/ecs_operator.py]: Regardless contrib files cant not be removed. In this specific case we must be backward compatible till Airflow 3

As for your question - in this file you need to support only classes that existed. We dont add new functionality

Comment (potiuk) [reply airflow/contrib/operators/ecs_operator.py]: I have an idea for the future. Should we move alll the deprecated/contrib operators from 1.10 out to a separate "apache-contrib" package (when we split out providers). 

That will allow to get rid of the contrib, and get rid of many 1.10 deprecations out of airlfow main repo at least (it can still be a dependency of Airflow package so that it is installed automatically when airflow is installed). @ashb  @kaxil WDYT?

**Review by vincbeck — APPROVED** [review]

File: airflow/contrib/operators/ecs_operator.py
Comment (ferruzzi) [reply airflow/contrib/operators/ecs_operator.py]: > As for your question - in this file you need to support only classes that existed. We dont add new functionality

I thought I understood this, but the CI is proving me wrong.  Before this CR, contrib looked like this:
```
from airflow.providers.amazon.aws.operators.ecs import ECSOperator, ECSProtocol

__all__ = ["ECSOperator", "ECSProtocol"]

warnings.warn(
    "This module is deprecated. Please use `airflow.providers.amazon.aws.operators.ecs`.",
    DeprecationWarning,
    stacklevel=2,
)
```
and `deprecated_classes.py` had this:
```
(
    'airflow.providers.amazon.aws.operators.ecs.ECSOperator',
    'airflow.contrib.operators.ecs_operator.ECSOperator',
),
```
ECSOperator and ECSProtocol were deprecated some time ago in favor of the names EcsOperator and EcsProtocol, so it was importing classes which were deprecated.  Do those deprecated classes still have to stay now as well?  It has been a full major release of the provider package so I wanted to start removing those old methods.  In this PR I am now renaming  EcsOperator to EcsRunTaskOperator, and EcsProtocol has been deprecated and moved into *.hooks.ecs.EcsProtocol

Comment (potiuk) [reply airflow/contrib/operators/ecs_operator.py]: Well. They shoudl be. But this is not a must actually.

Contrib were part of Airflow 1.10. When we moved to 2, we vowed to keep MOSTLY compatibility  with 1.10. I.e. people who still used 1.10 would be able to continue using them. There were a number of exceptions and back then we had "backport" providers so people could simply move to use backports "airfow.providers.*" but we stoppped (as planned) releasing those, so the "late-comers" will have to deal with conesequences ot potentially broken compatibility if they still use contrib.

Since then many of those Contrib operators and hooks were modified and they are not even backwards-compatible with the original Contrib ones. 

So technically speaking, if an operator is anyhow backwards incompatible we do not have to keep it in contrib. But if it largely is, it makes it easier for user to have smooth migration.

Comment (ferruzzi) [reply airflow/contrib/operators/ecs_operator.py]: @eladkal - @potiuk mentioned wanting to get a provider package release out and Amazon would be getting a major release, I'm hoping to get this in on that but I'm not really sure what needs to be done here now.  Any help would be appreciated.

Comment (eladkal) [reply airflow/contrib/operators/ecs_operator.py]: It will take me a few days to get to this

Comment (potiuk) [reply airflow/contrib/operators/ecs_operator.py]: I can take this one. But you need to rebase @ferruzzi - this one has problems from few days back when we fixed Flask 2.2 compatibility:

image

Comment (ferruzzi) [reply airflow/contrib/operators/ecs_operator.py]: ACK.   ~Will rebase and do a force-push.~ Done.

Thanks to both of you for the help.

Comment (ferruzzi) [reply airflow/contrib/operators/ecs_operator.py]: Done.

Comment (ferruzzi) [reply airflow/contrib/operators/ecs_operator.py]: I was a bit confused on how that chain of depredations should play out, but now with https://github.com/apache/airflow/pull/25543 I'm doubly confused on how the two will interact.

**Comment by ferruzzi** [comment]
Rebased and fixed (??) the deprecation notice.   This passes testing locally, let's see if it works.

**Comment by potiuk** [comment]
If the tests pass, do you think we should be fine with releasing it in this upcoming wave of providers?

**Comment by potiuk** [comment]
> If the tests pass, do you think we should be fine with releasing it in this upcoming wave of providers?

Ah yeah. I see the comment now :)

**Comment by eladkal** [comment]
> If the tests pass, do you think we should be fine with releasing it in this upcoming wave of providers?

I wont have time today to review the code.. i didnt do full review when I left the comment.

If the PR is OK from your side dont let me get in the way :)
I can chime in tommorow if my review is required.

**Comment by ferruzzi** [comment]
Alright, it just went green, so at least the CI is happy.  

@jarek - My main reason for wanting it in this release was because it's removing deprecated code.  Based on a chat with @eladkal, it doesn't sound like he thinks that's such a big deal.  If you both think that isn't a good enough reason to hold up the release then by all means push on without it.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Hmm. Some conflicts after merging #25543  - will you rebase @ferruzzi ?

**Comment by ferruzzi** [comment]
@potiuk Rebase done, passes all CI and static tests locally

File: airflow/providers/amazon/aws/operators/ecs.py
Comment (zachliu) [inline airflow/providers/amazon/aws/operators/ecs.py:55]: oh man, you passed all kwargs including the `aws_conn_id` to the BaseOperator
now 80% of my dags (i use EcsOperator heavily) are failing with
```
Broken DAG: [/usr/local/airflow/dags/my_dag.py] Traceback (most recent call last):
  File "/usr/local/airflow/.local/lib/python3.8/site-packages/airflow/models/baseoperator.py", line 390, in apply_defaults
    result = func(self, **kwargs, default_args=default_args)
  File "/usr/local/airflow/.local/lib/python3.8/site-packages/airflow/models/baseoperator.py", line 743, in __init__
    raise AirflowException(
airflow.exceptions.AirflowException: Invalid arguments were passed to EcsOperator (task_id: my-task). Invalid arguments were:
**kwargs: {'aws_conn_id': 'aws_default'}
``` 
is there a fix? will 2 simple pops do the job?

Comment (zachliu) [reply airflow/providers/amazon/aws/operators/ecs.py:55]: nvm, https://github.com/apache/airflow/pull/25989

**Pull request merged**
